{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6fdea5d-9418-412b-988c-fb22c19cb46c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -N 1 # number of nodes\n",
    "#SBATCH -n 16 # number of cores\n",
    "#SBATCH --mem=48G\n",
    "#SBATCH --job-name=\"Umar_preprocessing\"\n",
    "#SBATCH --mail-type=END\n",
    "#SBATCH --partition=computeq\n",
    "#SBATCH --mail-user=hrclndnn@memphis.edu\n",
    "#SBATCH -t 14-00:00:00\n",
    "#SBATCH -a 0-81\n",
    "\n",
    "module load bwa\n",
    "module load samtools\n",
    "module load picard\n",
    "module load sratoolkit\n",
    "module load gatk\n",
    "module load bcftools\n",
    "\n",
    "files=(SRR517013\n",
    "SRR518661\n",
    "SRR518662\n",
    "SRR518670\n",
    "SRR518671\n",
    "SRR518672\n",
    "SRR518673\n",
    "SRR518676\n",
    "SRR518677\n",
    "SRR518678\n",
    "SRR518679\n",
    "SRR518680\n",
    "SRR518684\n",
    "SRR518685\n",
    "SRR518686\n",
    "SRR518687\n",
    "SRR518688\n",
    "SRR518689\n",
    "SRR518699\n",
    "SRR518700\n",
    "SRR518701\n",
    "SRR518702\n",
    "SRR518703\n",
    "SRR827537\n",
    "SRR827574\n",
    "SRR827584\n",
    "SRR827585\n",
    "SRR827587\n",
    "SRR827588\n",
    "SRR827600\n",
    "SRR942197\n",
    "SRR942203\n",
    "SRR942211\n",
    "SRR942218\n",
    "SRR942235\n",
    "SRR942239\n",
    "SRR942240\n",
    "SRR942246\n",
    "SRR942248\n",
    "SRR942249\n",
    "SRR942251\n",
    "SRR942253\n",
    "SRR942260\n",
    "SRR942265\n",
    "SRR942266\n",
    "SRR942273\n",
    "SRR942275\n",
    "SRR942276\n",
    "SRR942277\n",
    "SRR942278\n",
    "SRR942279\n",
    "SRR942280\n",
    "SRR942281\n",
    "SRR942282\n",
    "SRR942283\n",
    "SRR942284\n",
    "SRR942285\n",
    "SRR942286\n",
    "SRR942287\n",
    "SRR942288\n",
    "SRR942289\n",
    "SRR942290\n",
    "SRR942291\n",
    "SRR942292\n",
    "SRR942293\n",
    "SRR942294\n",
    "SRR942295\n",
    "SRR942296\n",
    "SRR942297\n",
    "SRR942298\n",
    "SRR942299\n",
    "SRR942300\n",
    "SRR942301\n",
    "SRR942302\n",
    "SRR942303\n",
    "SRR942304\n",
    "SRR942305\n",
    "SRR942306\n",
    "SRR942307\n",
    "SRR942308\n",
    "SRR942309\n",
    "SRR942310\n",
    "SRR947747)\n",
    "\n",
    "file=${files[$SLURM_ARRAY_TASK_ID]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "172a8e14-38d7-4348-acd2-2336524e394e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#00 pull samples from NCBI\n",
    "cd /home/hrclndnn/sra_local/\n",
    "prefetch --max-size 100G ${file}\n",
    "fastq-dump --split-files ${file}\n",
    "\n",
    "\n",
    "#01 Align to reference genome\n",
    "bwa mem -M -t 16 /home/mdpllard/Genomes/Ursus_maritimus.fna /home/hrclndnn/WGSUmari/00_rethread/${file}_1.fastq /home/hrclndnn/WGSUmari/00_rethread/${file}_2.fastq > /home/hrclndnn/WGSUmari/01_mappedSam/S${file}_mapped.sam\n",
    "\n",
    "\n",
    "#02 Create a sorted bam file from the mapped sam file\n",
    "gatk --java-options '-Djava.io.tmpdir=/home/scratch/hrclndnn/' SortSam -I /home/hrclndnn/WGSUmari/01_mappedSam/${file}_mapped.sam -O /home/hrclndnn/WGSUmari/02_sortedBam/${file}.sorted.bam -SO coordinate \n",
    "\n",
    "#03 Mark duplicates\n",
    "gatk --java-options '-Djava.io.tmpdir=/home/scratch/hrclndnn/' MarkDuplicates -I /home/hrclndnn/WGSUmari/02_sortedBam/${file}.sorted.bam -O /home/hrclndnn/WGSUmari/03_markDups/${file}.dup.sorted.bam -M /home/hrclndnn/WGSUmari/03_markDups/${file}.dup.metrics --MAX_FILE_HANDLES 1000\n",
    "\n",
    "\n",
    "#04 Add or replace read groups\n",
    "cd /home/hrclndnn/WGSUmari/03_markDups/ #may not need, but sometimes GATK has errors related to paths\n",
    "gatk AddOrReplaceReadGroups -I /home/hrclndnn/WGSUmari/03_markDups/${file}.dup.sorted.bam -O /home/hrclndnn/WGSUmari/04_addRepGroups/${file}.RG.dup.sorted.bam --RGLB NCBI --RGPL illumina --RGPU NCBI --RGSM /home/hrclndnn/WGSUmari/04_addRepGroups/${file} \n",
    "\n",
    "#05 call variants\n",
    "samtools index /home/hrclndnn/WGSUmari/04_addRepGroups/${file}.RG.dup.sorted.bam\n",
    "gatk --java-options '-DGATK_STACKTRACE_ON_USER_EXCEPTION=true' HaplotypeCaller -R /home/mdpllard/Genomes/Ursus_maritimus.fna -I /home/hrclndnn/WGSUmari/04_addRepGroups/${file}.RG.dup.sorted.bam -O /home/hrclndnn/WGSUmari/05_callVariants/${file}.vcf.gz --heterozygosity 0.0005 -ERC GVCF\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "#changeRGSMnames\n",
    "#directory of scripts for each sample at: /home/hrclndnn/WGSUmari/scripts/changeRGSMnames\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f60d80e3-e068-4911-aaa9-183735d9b3bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "06_combinegVCFs.sh \n",
    "#!/bin/bash\n",
    "#SBATCH -N 1 # number of nodes\n",
    "#SBATCH -n 16 # number of cores\n",
    "#SBATCH --mem=48G\n",
    "#SBATCH --job-name=\"Ursus_maritimus_06\"\n",
    "#SBATCH --mail-type=END\n",
    "#SBATCH --partition=computeq\n",
    "#SBATCH --mail-user=hrclndnn@memphis.edu\n",
    "#SBATCH -t 14-00:00:00\n",
    "\n",
    "module load bwa\n",
    "module load samtools\n",
    "module load picard\n",
    "module load sratoolkit\n",
    "module load gatk\n",
    "module load bcftools\n",
    "\n",
    "#06_Combine gVCFs\n",
    "cd /home/hrclndnn/WGSUmari/05_callVariants/\n",
    "gatk CombineGVCFs -R /home/mdpllard/Genomes/Ursus_maritimus.fna \\\n",
    "\t-V SRR518661.vcf.gz \\\n",
    "\t-V SRR518662.vcf.gz \\\n",
    "\t-V SRR518670.vcf.gz \\\n",
    "\t-V SRR518671.vcf.gz \\\n",
    "\t-V SRR518672.vcf.gz \\\n",
    "\t-V SRR518673.vcf.gz \\\n",
    "\t-V SRR518676.vcf.gz \\\n",
    "\t-V SRR518677.vcf.gz \\\n",
    "\t-V SRR518678.vcf.gz \\\n",
    "\t-V SRR518679.vcf.gz \\\n",
    "\t-V SRR518680.vcf.gz \\\n",
    "\t-V SRR518684.vcf.gz \\\n",
    "\t-V SRR518685.vcf.gz \\\n",
    "\t-V SRR518686.vcf.gz \\\n",
    "\t-V SRR518687.vcf.gz \\\n",
    "\t-V SRR518688.vcf.gz \\\n",
    "\t-V SRR518689.vcf.gz \\\n",
    "\t-V SRR518699.vcf.gz \\\n",
    "\t-V SRR518700.vcf.gz \\\n",
    "\t-V SRR518701.vcf.gz \\\n",
    "\t-V SRR518702.vcf.gz \\\n",
    "\t-V SRR518703.vcf.gz \\\n",
    "\t-V SRR827537.vcf.gz \\\n",
    "\t-V SRR827574.vcf.gz \\\n",
    "\t-V SRR827584.vcf.gz \\\n",
    "\t-V SRR827585.vcf.gz \\\n",
    "\t-V SRR827587.vcf.gz \\\n",
    "\t-V SRR827588.vcf.gz \\\n",
    "\t-V SRR827600.vcf.gz \\\n",
    "\t-V SRR942197.vcf.gz \\\n",
    "\t-V SRR942203.vcf.gz \\\n",
    "\t-V SRR942211.vcf.gz \\\n",
    "\t-V SRR942213.vcf.gz \\\n",
    "\t-V SRR942218.vcf.gz \\\n",
    "\t-V SRR942235.vcf.gz \\\n",
    "\t-V SRR942239.vcf.gz \\\n",
    "\t-V SRR942240.vcf.gz \\\n",
    "\t-V SRR942246.vcf.gz \\\n",
    "\t-V SRR942248.vcf.gz \\\n",
    "\t-V SRR942249.vcf.gz \\\n",
    "\t-V SRR942251.vcf.gz \\\n",
    "\t-V SRR942253.vcf.gz \\\n",
    "\t-V SRR942260.vcf.gz \\\n",
    "\t-V SRR942263.vcf.gz \\\n",
    "\t-V SRR942265.vcf.gz \\\n",
    "\t-V SRR942266.vcf.gz \\\n",
    "\t-V SRR942273.vcf.gz \\\n",
    "\t-V SRR942275.vcf.gz \\\n",
    "\t-V SRR942276.vcf.gz \\\n",
    "\t-V SRR942277.vcf.gz \\\n",
    "\t-V SRR942278.vcf.gz \\\n",
    "\t-V SRR942279.vcf.gz \\\n",
    "\t-V SRR942280.vcf.gz \\\n",
    "\t-V SRR942281.vcf.gz \\\n",
    "\t-V SRR942282.vcf.gz \\\n",
    "\t-V SRR942283.vcf.gz \\\n",
    "\t-V SRR942284.vcf.gz \\\n",
    "\t-V SRR942285.vcf.gz \\\n",
    "\t-V SRR942286.vcf.gz \\\n",
    "\t-V SRR942287.vcf.gz \\\n",
    "\t-V SRR942288.vcf.gz \\\n",
    "\t-V SRR942289.vcf.gz \\\n",
    "\t-V SRR942290.vcf.gz \\\n",
    "\t-V SRR942291.vcf.gz \\\n",
    "\t-V SRR942292.vcf.gz \\\n",
    "\t-V SRR942293.vcf.gz \\\n",
    "\t-V SRR942294.vcf.gz \\\n",
    "\t-V SRR942295.vcf.gz \\\n",
    "\t-V SRR942296.vcf.gz \\\n",
    "\t-V SRR942297.vcf.gz \\\n",
    "\t-V SRR942298.vcf.gz \\\n",
    "\t-V SRR942299.vcf.gz \\\n",
    "\t-V SRR942300.vcf.gz \\\n",
    "\t-V SRR942301.vcf.gz \\\n",
    "\t-V SRR942302.vcf.gz \\\n",
    "\t-V SRR942303.vcf.gz \\\n",
    "\t-V SRR942304.vcf.gz \\\n",
    "\t-V SRR942305.vcf.gz \\\n",
    "\t-V SRR942306.vcf.gz \\\n",
    "\t-V SRR942307.vcf.gz \\\n",
    "\t-V SRR942308.vcf.gz \\\n",
    "\t-V SRR942309.vcf.gz \\\n",
    "\t-V SRR942310.vcf.gz \\\n",
    "\t-V SRR947747.vcf.gz \\\n",
    "\t-O /home/hrclndnn/WGSUmari/06_combineGVCFs/Umar.combined.Jun22.g.vcf.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d805b75c-2306-4f38-89f2-063027cbba42",
   "metadata": {},
   "outputs": [],
   "source": [
    "07_genoGVCFS.sh \n",
    "#!/bin/bash\n",
    "#SBATCH -N 1 # number of nodes\n",
    "#SBATCH -n 16 # number of cores\n",
    "#SBATCH --mem=48G\n",
    "#SBATCH --job-name=\"UmarPrep7\"\n",
    "#SBATCH --mail-type=END\n",
    "#SBATCH --partition=computeq\n",
    "#SBATCH --mail-user=hrclndnn@memphis.edu\n",
    "#SBATCH -t 14-00:00:00\n",
    "\n",
    "module load bwa\n",
    "module load samtools\n",
    "module load picard\n",
    "module load sratoolkit\n",
    "module load gatk\n",
    "module load bcftools\n",
    "\n",
    "#07_Genotype gVCFs\n",
    "gatk GenotypeGVCFs -R /home/mdpllard/Genomes/Ursus_maritimus.fna \\\n",
    "\t-V /home/hrclndnn/WGSUmari/06_combineGVCFs/Umar.combined.Jun1522.g.vcf.gz --heterozygosity 0.0005 \\\n",
    "\t-O /home/hrclndnn/WGSUmari/07_genotypeGVCFs/UmarGenotypedJun1522.vcf.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1085be20-81ec-4901-b822-2737a25cbe37",
   "metadata": {},
   "outputs": [],
   "source": [
    "#08 Filter VCFs\n",
    "#bcftools view \\\n",
    "#\t\t-i 'FMT/DP>4 & FMT/DP<300 && FMT/GQ>=20' \\ #filtering for depth of coverage and quality of calls\n",
    "##\t\t-s ^LA362 \\ #add a sample to process, or multiple samples as comma separated list\n",
    "#\t\t-m2 -M2 /home/hrclndnn/WGS_Ursus_maritimus/07_genotypeGVCFs/Umar_genotypedFeb1522.vcf.gz \\ #select only biallelic (excluding multiallelic) snps from the vcf.gz file from 07\n",
    "#\t\t-Oz -o /home/hrclndnn/WGS_Ursus_maritimus/08_bcftoolsFilter/Umar.filt.vcf.gz #outfile for 08\n",
    "bcftools view -i 'FMT/DP>4 & FMT/DP<300 && FMT/GQ>=20' -m2 -M2 /home/hrclndnn/WGSUmari/07_genotypeGVCFs/UmarGenotypedJun1522.vcf.gz \\\n",
    "\t-Oz -o /home/hrclndnn/WGSUmari/08_bcftoolsFilter/UmarJun1522.filt.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cf03e3e-dcd3-461b-ab88-c76955758078",
   "metadata": {},
   "outputs": [],
   "source": [
    "#09 filter SNPs in VCFs\n",
    "#bcftools view \\\n",
    "#\t\t-v snps /home/hrclndnn/WGS_Ursus_maritimus/08_bcftoolsFilter/Uamer.filt.vcf.gz \\ #select only biallelic snps from 08 outfile\n",
    "#\t\t-Oz -o /home/hrclndnn/WGS_Ursus_maritimus/09_filterSNPs/Uamer.snp.filt.vcf.gz   #create outfile for 09\n",
    "bcftools view -v snps /home/hrclndnn/WGSUmari/08_bcftoolsFilter/UmarJun1522.filt.vcf.gz -Oz -o /home/hrclndnn/WGSUmari/09filterSNPs/UmarJun1522.snp.filt.vcf.gz "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
